### ISSUE TITLE
Add posibility to retry queries to PostgreSQL for PostgreSQL database/table engine

### ISSUE THREAD

**Issue by StreamSilence** [ClickHouse/ClickHouse#62709]
If a query is executed after the Postgres database is restarted, a connection error occurs and the current clickhouse version is 23.10.4.25.
After I restarted the postgre database myself, I immediately executed the query sql and found that the results returned by two consecutive queries were connection errors, and the query was not successful until the third time.
There was already a similar issue before the discovery (Broken connection error after remote Postgres server restart when using remote table engine #33199)
Added a mechanism to retry the connection, the following figure is the exception log when the problem is reproduced, and it seems that there is no mechanism to re-obtain the connection to the PG database during the invocation process.

![clickhouse_001](https://github.com/ClickHouse/ClickHouse/assets/42864358/c9df1145-92fe-4058-98a4-e597b60a9829)

**Comment by kssenii** [comment]
> and it seems that there is no mechanism to re-obtain the connection to the PG database during the invocation process.

true, we use connection pool for postgresql connection. You can fix your issue by changing `postgresql_connection_pool_auto_close_connection` setting to not reuse the connections.

**Comment by acompagnon** [comment]
Hi,
 I have exactly the same issue and was looking at the setting `postgresql_connection_pool_auto_close_connection` you mentioned and have two questions about it :
 - In the [documentation](https://clickhouse.com/docs/en/operations/settings/settings#postgresql-connection-pool-auto-close-connection) it says the default value is true but when I see [Settings.h on master](https://github.com/ClickHouse/ClickHouse/blob/master/src/Core/Settings.h#L578) I see false as a default value
 - In my ClickHouse 21.12.3.32 when I make a select on system.settings table I don't have this setting, is it a setting that does not exist in this version ? Is the versioning of core settings is mentioned somewhere in the documentation ?
 
 Thanks a lot for your time

**Comment by kssenii** [comment]
> In the [documentation](https://clickhouse.com/docs/en/operations/settings/settings#postgresql-connection-pool-auto-close-connection) it says the default value is true but when I see [Settings.h on master](https://github.com/ClickHouse/ClickHouse/blob/master/src/Core/Settings.h#L578) I see false as a default value, is it wanted ?

it is a mistake, the doc is incorrect

> In my ClickHouse 21.12.3.32 when I make a select on system.settings table I don't have this setting, is it a setting that does not exist in this version ?

Yes.

>  Is the versioning of core settings is mentioned somewhere in the documentation ?

There is a system table `system.settings_changes`, it includes all the settings changes. But it was added in 22.3.
